python-checks/src/main/java/org/sonar/python/checks/InstanceMethodSelfAsFirstCheck.java

```diff
@@ -0,0 +1,197 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.check.RuleProperty;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.symbols.ClassSymbol;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.tree.AnyParameter;
+import org.sonar.plugins.python.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.ClassDef;
+import org.sonar.plugins.python.api.tree.Decorator;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.FunctionDef;
+import org.sonar.plugins.python.api.tree.HasSymbol;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.Parameter;
+import org.sonar.plugins.python.api.tree.ParameterList;
+import org.sonar.plugins.python.api.tree.RegularArgument;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key="S5720")
+public class InstanceMethodSelfAsFirstCheck extends PythonSubscriptionCheck {
+
+  // We allow "_" as sometimes it is conventionally used to signal that self won't be used.
+  private static final List<String> ALLOWED_NAMES = Arrays.asList("self", "_");
+  private static final List<String> EXCEPTIONS = Arrays.asList("__init_subclass__", "__class_getitem__", "__new__");
+
+  private static final String DEFAULT_IGNORED_DECORATORS = "classproperty";
+  private List<String> decoratorsToExclude;
+
+  @RuleProperty(
+    key = "ignoredDecorators",
+    description = "Comma-separated list of decorators which will disable this rule.",
+    defaultValue = DEFAULT_IGNORED_DECORATORS)
+  private String ignoredDecorators = DEFAULT_IGNORED_DECORATORS;
```

Comment:
I'm not sure about having this property.
I'd suggest ignoring all methods having a decorator. @nicolas-harraudeau-sonarsource wdyt?

Reply:
@andrea-guarino-sonarsource I was the one asking for this property.
On ITs we saw only one decorator: `@classproperty` from django. After some rapid research I saw that [some other project had an equivalent decorator](https://docs.astropy.org/en/stable/api/astropy.utils.decorators.classproperty.html) with the same name.

Having decorators on methods [is quite common](https://sourcegraph.com/search?q=%22%40%22+%22%28%22+%22%5Cn%22+%22def%22+%22%28self%22+lang:python+count:1000&patternType=regexp). Having a decorator which removes "self" or replaces it with something else should be quite rare. I don't expect devs to add more than 1 or 2 decorators, if they add any. Could we merge this and check on peach? If we see more than 2 decorators doing this we can reevaluate.

Reply:
ok, let's do that